nucypher/datastore/datastore.py

```diff
@@ -14,269 +14,222 @@
+            # no records for the query because lmdb orders the keys lexicographically.
+            # Ergo, if the current key doesn't match the query key, we know
+            # we have gone beyond the relevant keys and can `break` the loop.
+            # Additionally, if the record is already in the `valid_records`
+            # set (identified by the `record_id`, we call `continue`.
+            for db_key in db_cursor.iternext(keys=True, values=False):
+                curr_key = DatastoreKey.from_bytestring(db_key)
+                if not curr_key.compare_key(query_key):
+                    break
+                elif curr_key.record_id in valid_records:
+                    continue
+
+                record = partial(record_type, datastore_tx, curr_key.record_id)
+
+                # We pass the field to the filter_func if `filter_field` and
+                # `filter_func` are both provided. In the event that the
+                # given `filter_field` doesn't exist for the record or the
+                # `filter_func` returns `False`, we call `continue`.
+                if filter_field and filter_func:
+                    try:
+                        field = getattr(record(writeable=False), filter_field)
+                    except (TypeError, AttributeError):
+                        continue
+                    else:
+                        if not filter_func(field):
+                            continue
+
+                # If only a filter_func is given, we pass a readonly record to it.
+                # Likewise to the above, if `filter_func` returns `False`, we
+                # call `continue`.
+                elif filter_func:
+                    if not filter_func(record(writeable=False)):
+                        continue
+
+                # Finally, having a record that satisfies the above conditional
+                # constraints, we can add the record to the set
+                valid_records.add(record(writeable=writeable))
+
+            # If after the iteration we have no records, we raise `RecordNotFound`
+            if len(valid_records) == 0:
+                raise RecordNotFound(f"No records exist for the key from the specified query parameters: '{query_key}'")
+            # We begin the context manager try/finally block
+            try:
+                # At last, we yield the queried records
+                yield list(valid_records)
+            except (AttributeError, TypeError, DBWriteError) as tx_err:
+                # Handle `RecordNotFound` cases when `writeable` is `False`.
+                if not writeable and isinstance(tx_err, AttributeError):
+                    raise RecordNotFound(tx_err)
+                raise DatastoreTransactionError(f'An error was encountered during the transaction (no data was written): {tx_err}')
+            finally:
+                for record in valid_records:
+                    record.__dict__['_DatastoreRecord__writeable'] = False
```

Comment:
You have self in scope here - why do you need to directly mutate the object dict?

Reply:
I'm confused -- I'm not mutating `self` here, I'm mutating `record`.
This ensures that modifying attrs on the record, outside of the context manager, raises a `TypeError` and not fail silently.
Due to how `DatastoreRecord` attrs are used to access data, this avoids calling `__setattr__`.

Reply:
Ah - I must've misread it.  In any case this is a major code smell - whats going on here?

Reply:
> This ensures that modifying attrs on the record, outside of the context manager, raises a TypeError and not fail silently.
Due to how DatastoreRecord attrs are used to access data, this avoids calling __setattr__.

Sure - there sound design in there :+1:  - but mechanically this is awkward - I'm not entirely sure what the code is telling us here, perhaps that the scope is not correct :thinking:

Reply:
So if we have a query:
```
with datastore.query_by(FooRecord, writeable=True) as my_foo_records:
    my_record = my_foo_records[0]
    .... # Do something with my_record

# I later try to do something with `my_record` when I'm not supposed to:
my_record.foobar = b'uh oh'
```

I expect the `my_record.foobar` to be an error -- it doesn't have access to a transaction to even set the `foobar` field, but I don't want it to even get that far. This should be an error caught in `__setattr__`.

Reply:
To be totally clear - you are trying to access a protected attribute in another class, which is intentional obscured.  

A fix here might be as simple as loosening up and just removing a single underscore :-)

Reply:
```suggestion
                    record._writeable(False)
```

Reply:
I guess it could be written as `_writeable`, but that also has implications on `DatastoreRecord` as well. Presently, field types are expressed as `_<name>`, so this would make `_writeable` share the same "namespace" as those types, which was my reasoning for making this `__writeable` since I know nothing outside of the `Datastore` would (or should) be accessing it.